### PR TITLE
ConnectionService: adding new table-specific connection mgmt

### PR DESCRIPTION
**Pull request by bryanlatten**
The existing contract guarantees DB replication lag safety, by enforcing that reads come from master *after* an existing write. But...this write-then-read policy is *only* valid on an individual table basis. A write to table `ABC` will not affect the replication lag for `DEF`.

- [x] moved PDO-specific commands out of AbstractAdapter into PdoAdapter, including tests
- [x] incorporate `get*` style commands: moved logic from fetch to `extract` fxs, used `queryTable` instead of `query`
- [x] event handler: pre-query has master usage intent, post-query has master *actual* usage
- [x] close connection reset
- [x] consolidated query generation
- do anything different on reconnect? `probably not`
- do anything different on transactions? `probably not`
- check into usage of quote() `probably not`

screen shot 2017-03-21 at 3 28 53 pm

### CODE REVIEW

**Comment by ingluisjimenez** [comment]
👍

**Comment by jimdelois** [comment]
@bryanlatten cc: @ingluisjimenez I assume this is intended to address https://github.com/adobe-community/issues/issues/14171 so it can be closed out, correct?

**Comment by bryanlatten** [comment]
@jimdelois yes, the data model does not have enough insight to make those decisions (deliberately). The existing architecture of delaying the connection capture until the last second allows this transformation to take place under the hood, almost entirely! unlike propel and doctrine, which do the connection retrieval upfront

File: src/AdapterAbstract.php [hunk 1]
```diff
@@ -58,228 +57,125 @@ public function __construct( ConnectionService $connection, EventDispatcherInter
   /**
    * {@inheritDoc}
    */
-  abstract public function update( $table, array $data, $where );
+  abstract public function fetchOne( $sql, array $parameters = null, $master = false );
```
Comment (bryanlatten) [inline src/AdapterAbstract.php]: many PDO-specific implementations were pushed down to the PDO level, and made abstract

File: src/AdapterAbstract.php [hunk 2]
```diff
@@ -295,52 +191,41 @@ public function getCol( $table, $column, $where, $master = false ) {
   /**
    * {@inheritDoc}
    */
-  public function getColumn( $table, $column, $where, $master = false ) {
-
-    list( $where_sql, $where_prepared ) = $this->_buildWhere( $where );
+  public function query( $sql, array $parameters = null, $use_master = false ) {
```
Comment (bryanlatten) [inline src/AdapterAbstract.php]: now that `_execute` has been promoted to a protected abstract, can promote the implementation of `query` (and others) to the abstract

File: src/AdapterAbstract.php [hunk 3]
```diff
@@ -364,21 +249,39 @@ public function closeConnection() {
...
+   *
+   * @param string|null $table if known, will be used to further segment connection handling
+   * @param string $sql        query to prepare and execute
+   * @param array  $parameters data to be used in $sql
+   * @param bool   $use_master whether to use write or read connection for statement
+   * @param int    $retries    number of attempts executed on $statement, used to prevent infinite recursion, one retry is max
+   *
+   * @return mixed
+   */
+  abstract protected function _execute( $table, $sql, array $parameters = null, $use_master = false, $retries = 0 );
```
Comment (bryanlatten) [inline src/AdapterAbstract.php]: promoted out of PDO

File: src/AdapterInterface.php [hunk 4]
```diff
@@ -119,6 +119,38 @@ public function queryMaster( $sql, array $parameters = null );
...
+   * default to helper methods to avoid writing raw SQL
+   *
+   * @param string $table       explicitly call out the table in use
+   * @param string $sql         can contain prepared statement placeholders ('?'' or ':key')
+   * @param array  $parameters  format must match placeholders, if using ?, a flat array, otherwise key/value
+   * @param bool   $use_master  flags to use a master or replica connection (subject to connection rules)
+   *
+   * @return PDOStatement  post-execution statement
+   */
+  public function queryTableMaster( $table, $sql, array $parameters = null );
```
Comment (bryanlatten) [inline src/AdapterInterface.php]: new `query` interface, with table upfront. this is instead of crappily adding it to the back of the param stack

File: docker-compose.yml [hunk 5]
```diff
@@ -5,7 +5,7 @@ web:
   volumes:
    - ./:/app
   ports:
-  - '80'
+  - '8080'
```
Comment (bryanlatten) [inline docker-compose.yml]: now you can properly create code coverage on the compose stack AND view it within the container!

File: src/ConnectionService.php [hunk 6]
```diff
@@ -48,11 +63,25 @@ public function __construct( ConfigService $config ) {
 
 
   /**
+   * @param string $table  [NEW] follow connection management rules PER-TABLE when supplied
+   *
    * @return PDO
    */
-  public function getMaster() {
+  public function getMaster( $table = null ) {
```
Comment (bryanlatten) [inline src/ConnectionService.php]: <<HERE>> this is the majority of the new logic!

File: src/ConnectionService.php [hunk 7]
```diff
@@ -48,11 +63,25 @@ public function __construct( ConfigService $config ) {
...
+
+    if ( !$this->_master_flag ) {
 
-    if ( empty( $this->_master ) ) {
+      if ( $table ) {
+        // NOTE: using junky array index to store table name to prevent having to search a flat array before entering a value.
```
Comment (be-dmitry) [inline src/ConnectionService.php]: I'd rephrase with less snark here. Remember this is an open source lib

File: src/ConnectionService.php [hunk 8]
```diff
@@ -36,6 +40,17 @@ class ConnectionService {
    */
   private $_config;
 
+  /**
+   * @var string[] a list of any tables that have used the master connection
+   */
+  private $_master_tables = [];
+
+  /**
+   * Enabled to trigger fallback to table-inspecific connection behavior (legacy)
```
Comment (be-dmitry) [inline src/ConnectionService.php]: inspecific isn't really a word. condier `table-agnostic`

File: src/ConnectionService.php (hunk 7 above)
Comment (bryanlatten) [inline src/ConnectionService.php]: is that a snarky comment? a junky array index?

File: src/AdapterInterface.php [hunk 9]
```diff
@@ -119,6 +119,38 @@ public function queryMaster( $sql, array $parameters = null );
...
+   * Prepares and executes a raw SQL statement, with explicit table usage
+   *
+   *
+   * IMPORTANT: provided for the thinnest SQL compatibility possible,
+   * default to helper methods to avoid writing raw SQL
+   *
+   * @param string $table       explicitly call out the table in use
+   * @param string $sql         can contain prepared statement placeholders ('?'' or ':key')
+   * @param array  $parameters  format must match placeholders, if using ?, a flat array, otherwise key/value
+   * @param bool   $use_master  flags to use a master or replica connection (subject to connection rules)
```
Comment (bryanlatten) [inline src/AdapterInterface.php]: actually, it was a missing `$use_master` parameter, whoops!

File: README.md [hunk 10]
```diff
@@ -132,6 +133,20 @@ Integration testing: leveraging Docker, using actual mysql container
...
 <tr>
+<td>queryTable</td>
+<td>$adapter->queryTable( ‘table’, "SELECT * FROM `table` WHERE id=? AND enabled=?", [ 12345, 0 ] );</td>
+<td>PDOStatement</td>
+<td>*PDOStatement is already executed</td>
+</tr>
+
+<tr>
+<td>queryTableMaster</td>
+<td>$adapter->queryMaster( ‘table’, "SELECT * FROM `table` WHERE id=:id AND enabled=:enabled, [ ':id' => 12345, ':enabled' => 0 ] );</td>
```
Comment (ingluisjimenez) [inline README.md]: typo `$adapter->queryTableMaster(`

**Comment by be-dmitry** [comment]
👍

**Comment by TJP85** [comment]
👍

File: src/Adapters/PdoAdapter.php [hunk 11]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
-   * @return PDOStatement
+   * @return array  [ 0 => string: SQL, 1 => array: positional parameters ]
+   */
+  public function prepStarQuery( $table, $where ) {
+
+    list( $where_sql, $where_prepared ) = $this->_buildWhere( $where );
+
+    $sql = sprintf( "SELECT * FROM %s %s", $this->_quoteTable( $table ), $where_sql );
+
+    $parameters = ( empty( $where_prepared ) )
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `$parameters = $where_prepared ?: null;` no?

File: src/Adapters/PdoAdapter.php [hunk 12]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+   *
+   * @return array  [ 0 => string: SQL, 1 => array: positional parameters ]
+   */
+  public function prepSelectorQuery( $table, $field, $where ) {
+
+    list( $where_sql, $where_prepared ) = $this->_buildWhere( $where );
+
+    $sql = sprintf( "SELECT %s FROM %s %s", $this->_quoteColumn( $field ), $this->_quoteTable( $table ), $where_sql );
+
+    $parameters = ( empty( $where_prepared ) )
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `$parameters = $where_prepared ?: null;` no?

File: src/Adapters/PdoAdapter.php [hunk 13]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+
+  } // prepSelectorQuery
+
+
+  /**
+   * NOTE: use Traversable as a stand-in for typehinting that does not work properly for testing mocks
+   *
+   * @param PDOStatement $statement
+   *
+   * @return mixed
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `null|mixed`

File: src/Adapters/PdoAdapter.php [hunk 14]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+
+  /**
+   * NOTE: use Traversable as a stand-in for typehinting that does not work properly for testing mocks
+   *
+   * @param PDOStatement $statement
+   *
+   * @return array
    */
-  protected function _execute( $sql, array $parameters = null, $use_master = false, $retries = 0 ) {
+  public function extractAssocValues( \Traversable $statement ) {
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `\PDOStatement $statement`

File: src/Adapters/PdoAdapter.php [hunk 15]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+
+
+  /**
+   * NOTE: use Traversable as a stand-in for typehinting that does not work properly for testing mocks
+   *
+   * @param PDOStatement $statement
+   *
+   * @return array
+   */
+  public function extractArrayValues( \Traversable $statement ) {
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `\PDOStatement $statement`

File: src/Adapters/PdoAdapter.php [hunk 16]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+
+
+  /**
+   * NOTE: use Traversable as a stand-in for typehinting that does not work properly for testing mocks
+   *
+   * @param PDOStatement $statement
+   *
+   * @return array
+   */
+  public function extractKeyValues( \Traversable $statement ) {
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `\PDOStatement $statement`

File: src/Adapters/PdoAdapter.php [hunk 17]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+
+
+  /**
+   * NOTE: use Traversable as a stand-in for typehinting that does not work properly for testing mocks
+   *
+   * @param PDOStatement $statement
+   *
+   * @return mixed
+   */
+  public function extractOneValue( \Traversable $statement ) {
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `\PDOStatement $statement`

File: src/Adapters/PdoAdapter.php [hunk 18]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+
+
+  /**
+   * NOTE: use Traversable as a stand-in for typehinting that does not work properly for testing mocks
+   *
+   * @param PDOStatement $statement
+   *
+   * @return array
+   */
+  public function extractColumnAssocValues( \Traversable $statement ) {
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `\PDOStatement $statement`

File: src/Adapters/PdoAdapter.php [hunk 19]
```diff
@@ -244,56 +368,243 @@ public function closeConnection() {
...
+
+
+  /**
+   * NOTE: use Traversable as a stand-in for typehinting that does not work properly for testing mocks
+   *
+   * @param PDOStatement $statement
+   *
+   * @return array
+   */
+  public function extractPairValues( \Traversable $statement ) {
```
Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: `\PDOStatement $statement`

**Comment by ingluisjimenez** [comment]
👍

File: src/Adapters/PdoAdapter.php (hunk 17 above)
Comment (bryanlatten) [inline src/Adapters/PdoAdapter.php]: you missed the comment. `PDOStatement` is not cleanly mockable, so Traversable is being used to typehint "something." Same comment for all of the below. I wouldn't randomly typehint traversable for no reason :)

Comment (ingluisjimenez) [inline src/Adapters/PdoAdapter.php]: but you are not using any feature from `\Traversable`, why typehinting at all then? seems like more of a confusion than any benefit.

Comment (bryanlatten) [inline src/Adapters/PdoAdapter.php]: hence the comment. if you don't read the comment, you get nothing but confusion.
